internal/cortex/frontend/transport/utils.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) The Cortex Authors.
+// Licensed under the Apache License 2.0.
```

Comment:
Move this file to a sub-directory and a different package. This package shouldn't see `http.Request` to decouple with the handler.
internal/cortex/frontend/transport/utils/utils.go
```
package utils

// Handler holds an instance of FailedQueryCache and calls its methods
type FailedQueryCache struct {
        cacheableResponseCodes ...
	regex               ...
	errorExtract  ...
}

func NewFailedQueryCache() {
  // Initial the internal fields
}

func (f *FailedQueryCache) UpdateFailedQueryCache(err error, queryExpressionNormalized string, queryExpressionRangeLength int) bool, string {

  // return true if the query is cached so that the call site can bump a counter.
  // return a message as a string so that the call site can log the outcome.
}

func (f *FailedQueryCache) QueryHitCache(...) bool, string {
...
}
```
Add a unit test file internal/cortex/frontend/transport/utils/utils_test.go

Paste utils.go and ask ChatGPT to write a unit test file.